Signal/src/network/GiphyDownloader.swift

```diff
@@ -0,0 +1,410 @@
+        DispatchQueue.main.async {
+            self.assetMap.set(key:assetRequest.rendition.url, value:asset)
+            self.activeAssetRequests.remove(assetRequest)
+            assetRequest.requestDidSucceed(asset:asset)
+            self.startRequestIfNecessary()
+        }
+    }
+
+    private func assetRequestDidFail(assetRequest: GiphyAssetRequest) {
+        DispatchQueue.main.async {
+            self.activeAssetRequests.remove(assetRequest)
+            assetRequest.requestDidFail()
+            self.startRequestIfNecessary()
+        }
+    }
+
+    private func startRequestIfNecessary() {
+        AssertIsOnMainThread()
+
+        DispatchQueue.main.async {
+            guard self.activeAssetRequests.count < self.kMaxAssetRequestCount else {
+                return
+            }
+            guard let assetRequest = self.popNextAssetRequest() else {
+                return
+            }
+            guard !assetRequest.wasCancelled else {
+                // Discard the cancelled asset request and try again.
+                self.startRequestIfNecessary()
+                return
+            }
+            guard UIApplication.shared.applicationState == .active else {
+                // If app is not active, fail the asset request.
+                self.assetRequestDidFail(assetRequest:assetRequest)
+                self.startRequestIfNecessary()
+                return
+            }
+
+            self.activeAssetRequests.insert(assetRequest)
+
+            if let asset = self.assetMap.get(key:assetRequest.rendition.url) {
+                // Deferred cache hit, avoids re-downloading assets that were
+                // downloaded while this request was queued.
+
+                self.assetRequestDidSucceed(assetRequest : assetRequest, asset: asset)
+                return
+            }
+
+            guard let downloadSession = self.giphyDownloadSession() else {
+                owsFail("\(GiphyDownloader.TAG) Couldn't create session manager.")
+                self.assetRequestDidFail(assetRequest:assetRequest)
+                return
+            }
+
+            // Start a download task.
+            let task = downloadSession.downloadTask(with:assetRequest.rendition.url as URL)
+            task.assetRequest = assetRequest
+            task.resume()
+        }
+    }
+
+    private func popNextAssetRequest() -> GiphyAssetRequest? {
+        AssertIsOnMainThread()
+
+        var activeAssetRequestURLs = Set<NSURL>()
+        for assetRequest in activeAssetRequests {
+            activeAssetRequestURLs.insert(assetRequest.rendition.url)
+        }
+
+        // Prefer the first "high" priority request;
+        // fall back to the first "low" priority request.
+        for priority in [GiphyRequestPriority.high, GiphyRequestPriority.low] {
+            for (assetRequestIndex, assetRequest) in assetRequestQueue.enumerated() where assetRequest.priority == priority {
```

Comment:
TIL about where.

Reply:
As far as I can tell Swift is all about this kind of syntactic sugar.